napari/plugins/io.py

```diff
@@ -76,3 +78,191 @@ def read_data_with_plugins(
+    plugin_name : str, optional
+        Name of the plugin to use for saving. If None then all plugins
+        corresponding to appropriate hook specification will be looped
+        through to find the first one that can save the data.
+    plugin_manager : plugins.PluginManager, optional
+        Instance of a napari PluginManager.  by default the main napari
+        plugin_manager will be used.
+
+    Returns
+    -------
+    bool
+        Return True if data is successfully written.
+    """
+    if isinstance(layers, list):
+        if len(layers) > 1:
+            return _write_multiple_layers_with_plugins(
+                path, layers, plugin_name, plugin_manager
+            )
+        elif len(layers) == 1:
+            layers = layers[0]
+    if isinstance(layers, Layer):
+        return _write_single_layer_with_plugins(
+            path, layers, plugin_name, plugin_manager
+        )
+    return False
+
+
+def _write_multiple_layers_with_plugins(
+    path: str,
+    layers: List[Layer],
+    plugin_name: Optional[str] = None,
+    plugin_manager=napari_plugin_manager,
+):
+    """Write data from multiple layers data with a plugin.
+
+    If a ``plugin_name`` is not provided we loops through plugins to find the
+    first one that knows how to handle the combination of layers and is able to
+    write the file. If no plugins offer ``napari_get_writer`` for that
+    combination of layers then the default ``napari_get_writer`` will create a
+    folder and call ``napari_write_<layer>`` for each layer using the
+    ``layer.name`` variable to modify the path such that the layers are written
+    to unique files in the folder.
+
+    If a ``plugin_name`` is provided, then we call we call
+    ``napari_get_writer`` for that plugin, and if it doesn’t return a
+    WriterFunction we error, otherwise we call it and if that fails if it we
+    error.
+
+    Exceptions will be caught and stored as PluginErrors
+    (in plugins.exceptions.PLUGIN_ERRORS)
+
+    Parameters
+    ----------
+    path : str
+        The path (file, directory, url) to write.
+    layers : List of napari.layers.Layer
+        List of napari layers to write.
+    plugin_manager : plugins.PluginManager, optional
+        Instance of a napari PluginManager.  by default the main napari
+        plugin_manager will be used.
+    """
+    layer_data = [layer.as_layer_data_tuple() for layer in layers]
+    layer_types = [ld[2] for ld in layer_data]
+
+    hook_caller = plugin_manager.hook.napari_get_writer
+    if plugin_name:
+        # if plugin has been specified we just directly call napari_get_writer
+        # with that plugin_name.
+        implementation = hook_caller.get_plugin_implementation(plugin_name)
+        writer_function = hook_caller(
+            _plugin=plugin_name, path=path, layer_types=layer_types
+        )
+    else:
+        (writer_function, implementation) = hook_caller(
+            path=path, layer_types=layer_types, _return_impl=True
```

Comment:
@tlambert03 I'm running into the following error here:

```python-traceback
viewer.save_layers('njs-test', selected=False)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-ccf374d6cb14> in <module>
----> 1 viewer.save_layers('njs-test', selected=False)

~/GitHub/napari/napari/components/add_layers_mixin.py in save_layers(self, path, selected, plugin)
   1001             return
   1002 
-> 1003         write_layers_with_plugins(path=path, layers=layers, plugin_name=plugin)
   1004 
   1005 

~/GitHub/napari/napari/plugins/io.py in write_layers_with_plugins(path, layers, plugin_name, plugin_manager)
    140         if len(layers) > 1:
    141             return _write_multiple_layers_with_plugins(
--> 142                 path, layers, plugin_name, plugin_manager
    143             )
    144         elif len(layers) == 1:

~/GitHub/napari/napari/plugins/io.py in _write_multiple_layers_with_plugins(path, layers, plugin_name, plugin_manager)
    198     else:
    199         (writer_function, implementation) = hook_caller(
--> 200             path=path, layer_types=layer_types, _return_impl=True
    201         )
    202 

TypeError: cannot unpack non-iterable function object
```

Reply:
fixed.  (The `_return_impls` argument is gone after #1153.). I unskipped your tests and they all pass now